dataland-backend/src/main/kotlin/org/dataland/datalandbackend/services/PrivateDataManager.kt

```diff
@@ -301,4 +268,64 @@ class PrivateDataManager(
         return documentInMemoryStorage[hash]
     }
+
+    /**
+     * Retrieves a private sme data object from the private storage
+     * @param dataId the dataId of the dataset to be retrieved
+     * @param correlationId the correlationId of the request
+     */
+    fun getPrivateDataSet(dataId: String, correlationId: String): StorableDataSet {
+        return dataManagerUtils.getDataSet(
+            dataId, DataType.of(SmeData::class.java), correlationId,
+            ::getDataFromCacheOrStorageService,
+        )
+    }
+    private fun getDataFromCacheOrStorageService(dataId: String, correlationId: String): String {
+        return jsonDataInMemoryStorage[dataId] ?: dataManagerUtils
+            .getDataFromStorageService(
+                dataId, correlationId,
+                ::getPrivateData,
+            )
+    }
+    private fun getPrivateData(dataId: String, correlationId: String): String {
+        return storageClientUtils.getDataByIdFromStorageClient(dataId, correlationId)
+    }
+
+    /**
+     * This method retrieves a document from the storage
+     * @param dataId the dataId connected to document to be retrieved
+     * @param hash the hash of the requested document
+     * @param correlationId the correlationId of the request
+     */
+    @Suppress("TooGenericExceptionCaught")
+    fun retrievePrivateDocumentById(dataId: String, hash: String, correlationId: String): DocumentStream {
+        // TODO Wenn man nach einem document sucht, was es nicht gibt, entsteht hier aktuell ein internal server error
+        // TODO Grund: Die nächste Zeile kriegt eine out of bounds exception.  Das müssen wir iwie anders handlen,
+        // TODO Vorschlag: Passende exception schmeißen (ResourceNotFoundException z.B.)
+        val eurodatId: String
+        try {
+            eurodatId = dataIdToHashMappingRepository.findByDataIdAndHash(dataId, hash)[0].eurodatId
+        } catch (e: IndexOutOfBoundsException) {
+            logger.info("Dataset with id $dataId could not be found. Correlation ID: $correlationId")
+            throw ResourceNotFoundApiException(
+                "Document not found",
+                "No document for dataId $dataId with the hash: $dataId could be found in the data store.",
+                e,
+            )
+        }
+        val inMemoryStoredDocument = documentInMemoryStorage[hash]
+        return if (inMemoryStoredDocument != null) {
+            logger.info("Received document for eurodatId $eurodatId from temporary storage")
+            DocumentStream(hash, DocumentType.Pdf, InputStreamResource(ByteArrayInputStream(inMemoryStoredDocument)))
+        } else {
+            logger.info("Received document for eurodatId $eurodatId from storage service")
+            DocumentStream(
+                hash, DocumentType.Pdf,
+                InputStreamResource(
+                    storageClientUtils
+                        .getBlobFromExternalStorageClient(eurodatId, correlationId),
+                ),
```

Comment:
Ich fände die logs wertvoller, wenn wir über dem "val inMemoryStoreDocument..." einmal loggen, mit welcher eurodatId wir die dataId und den hash gemappt haben.
Sonst verwirrt den darauffolgende Log vllt ein wenig, weil man sich fragt, woher die Info der eurodatId jetzt plötzlich herkommt.

Reply:
Eingefügt

Reply:
Passt noch nicht:

Da steht jetzt 

"Using dataId $dataId and hash $hash to determine necessary eurodatId of the document. " +
                "CorrelationId $correlationId"
                
Aber das hat er zu diesem Zeitpunkt schon längst getan.
Sollte eher loggen:

"Matched dataId X and hash Y with eurodatId Z. CorrelationId: asölfkjaös"

Reply:
Geändert